woocommerce/woocommerce-gateway-stripe
# Add order meta to track UPE processed orders

# Changes proposed in this Pull Request:

During testing it was discovered that refreshing the order received page when using UPE could create an issue where the order is reprocessed when it was marked as failed. The exact scenario was a test using a SEPA payment method that immediately creates a dispute for the charge. The dispute would come in and the order would be marked `failed` appropriately. However, if you refreshed the order received page after this, the order would be moved back to `processing`.

This PR addresses this issue as well as any others of this nature by making sure that refreshing the order received page does not reprocess the order. This is accomplished by using an order meta field to track once an order has been processed and to return early if it has already been processed.

One question I was unsure of is when to properly set the order meta as processed. I opted to set it at the very end so that if there was an issue encountered earlier in the process, there is still a chance to try again with a refresh. I'd be open to moving this to another spot during the flow.

# Testing instructions

This test will rely on webhooks, so be sure to set up the webhook listener:

- Install [Stripe CLI](https://stripe.com/docs/stripe-cli).
- Run `stripe login` and select the Stripe account you're connected to.
- Run `stripe listen --forward-to 'http://localhost:8082/?wc-api=wc_stripe'` and copy the webhook signing secret (`whsec_******`) to the plugin settings.

Once the listener is running, run the following test:

- With this branch pulled down and UPE enabled with at least the SEPA payment method enabled within UPE.
- Add a product to the cart and go to checkout.
- Confirm the UPE displays and SEPA is an option.
- Select SEPA and use the following IBAN to checkout: `AT591904300235473203` this will succeed and then immediately create a dispute.
- Once on the order received page, do not close it!
- As merchant, go to check on the order status and it should be marked failed. I have found the webhooks to be a little unstable at times and so if they don't fire try restarting the listener and rerunning the test.
- Now refresh the order received page as the customer.
- As the merchant, confirm that the order is still marked as failed.
 
In addition to this specific test, multiple scenarios could be arranged by checking out and keeping the order received page up, then updating the order status as merchant and refreshing the order received page to be sure the status is not updated again.


-------------------
